src/artm/score/perplexity.cc

```diff
@@ -163,17 +191,70 @@ void Perplexity::AppendScore(const Score& score, Score* target) {
     BOOST_THROW_EXCEPTION(::artm::core::InternalError(error_message));
   }
 
-  perplexity_target->set_normalizer(perplexity_target->normalizer() +
-                                    perplexity_score->normalizer());
-  perplexity_target->set_raw(perplexity_target->raw() +
-                             perplexity_score->raw());
-  perplexity_target->set_zero_words(perplexity_target->zero_words() +
-                                    perplexity_score->zero_words());
-  perplexity_target->set_value(exp(- perplexity_target->raw() / perplexity_target->normalizer()));
-
-  VLOG(1) << "normalizer=" << perplexity_target->normalizer()
-          << ", raw=" << perplexity_target->raw()
-          << ", zero_words=" << perplexity_target->zero_words();
+  bool empty_target = !perplexity_target->class_id_info_size() && !perplexity_target->normalizer();
+  bool score_has_class_ids = perplexity_score->class_id_info_size();
+  bool target_has_class_ids = empty_target ? score_has_class_ids : perplexity_target->class_id_info_size();
+  if (target_has_class_ids != score_has_class_ids) {
+    std::stringstream ss;
+    ss <<"Inconsistent new content of perplexity score. Old content uses class ids: " << target_has_class_ids;
+    BOOST_THROW_EXCEPTION(::artm::core::InternalError(ss.str()));
+  }
+
+  double pre_value = 0.0;
+  if (target_has_class_ids) {
+    for (size_t i = 0; i < perplexity_score->class_id_info_size(); ++i) {
+      auto src = perplexity_score->class_id_info(i);
+
+      bool was_added = false;
+      for (size_t j = 0; j < perplexity_target->class_id_info_size(); ++j) {
+        if (perplexity_score->class_id_info(i).class_id() == perplexity_target->class_id_info(j).class_id()) {
+          // update existing class_id info
+          auto dst = perplexity_target->mutable_class_id_info(j);
+          dst->set_normalizer(dst->normalizer() + src.normalizer());
+          dst->set_raw(dst->raw() + src.raw());
+          dst->set_zero_words(dst->zero_words() + src.zero_words());
+
+          was_added = true;
+          break;
+        }
+      }
+
+      if (!was_added) {
+        // add new class_id info
+        auto dst = perplexity_target->add_class_id_info();
+        dst->set_class_id(src.class_id());
+        dst->set_normalizer(src.normalizer());
+        dst->set_raw(src.raw());
+        dst->set_zero_words(src.zero_words());
+      }
+    }
+
+    for (size_t j = 0; j < perplexity_target->class_id_info_size(); ++j) {
+      auto score = perplexity_target->class_id_info(j);
+      pre_value += score.raw() / score.normalizer();
+
```

Comment:
@MelLain Should we weigh log-likelihoods of different modalities according to their `class_weight` values?